corehq/apps/locations/permissions.py

```diff
@@ -20,34 +21,60 @@ def is_locations_admin(view_fn):
     return locations_access_required(domain_admin_required(view_fn))
 
 
-def user_can_edit_location(user, location, project):
-    """
-    Expects SQLLocation
-    """
-    if not project.location_restriction_for_users:
+@quickcache(['user._id', 'project.name'])
+def editable_locations_ids(user, project):
+    if (user.is_domain_admin(project.name) or
+            not project.location_restriction_for_users):
+        return (SQLLocation.by_domain(project.name)
+                           .values_list('location_id', flat=True))
+
+    user_loc = user.get_location(project.name)
+    if not user_loc:
+        return []
+
+    return list(user_loc.sql_location.get_descendants(include_self=True)
+                                 .values_list('location_id', flat=True))
+
+
+def user_can_edit_location(user, sql_location, project):
+    if (user.is_domain_admin(project.name) or
+            not project.location_restriction_for_users):
         return True
 
-    user_loc = user.get_location(location.domain)
+    user_loc = user.get_location(sql_location.domain)
     if user_loc:
         user_loc = user_loc.sql_location
-    return user_loc is None or user_loc.is_direct_ancestor_of(location)
+    return user_loc is None or user_loc.is_direct_ancestor_of(sql_location)
 
 
-def user_can_view_location(user, location, project):
-    """
-    Expects SQLLocation
-    """
-    if not project.location_restriction_for_users:
+def viewable_locations_ids(user, project):
+    if (user.is_domain_admin(project.name) or
+            not project.location_restriction_for_users):
+        return (SQLLocation.by_domain(project.name)
+                           .values_list('location_id', flat=True))
+
+    user_loc = user.get_location(project.name)
+    if not user_loc:
+        return []
+
+    return (list(user_loc.sql_location.get_ancestors()
+            .values_list('location_id', flat=True)) +
+            editable_locations_ids(user, project))
```

Comment:
That's probably correct, as both of those are instancemethods which return querysets - not model manager methods.  This seems fine though, as it's only 2 calls (rather than n)